{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contact: Chao Ning\n",
    "ningchao(at)sdau(dot)edu(dot)cn  \n",
    "ningchao91(at)gmail(dot)com  \n",
    "\n",
    "# Install\n",
    "\n",
    "## Dependencies\n",
    "* numpy>=1.16.0  \n",
    "* pandas>=0.19.0  \n",
    "* scipy>=1.1.1  \n",
    "* cffi>=1.12.0  \n",
    "* pandas_plink>=2.0.0  \n",
    "* tqdm>=4.43.0  \n",
    "\n",
    "We recommend using a Python distribution such as [Anaconda](https://www.anaconda.com/distribution/) (Python 3.7 version). This distribution can be used on Linux and Windows and is free. It is the easiest way to get all the required package dependencies. \n",
    "\n",
    "## Quick install\n",
    "\\> pip install gmat\n",
    "## Detailed Package Install Instructions:\n",
    "(1) Install the dependent packages  \n",
    "(2) Go to the directory of GMAT and type  \n",
    "\\> python setup.py install"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# REMMAX function  \n",
    "Rapid Epistatic Mixed Model Association Studies\n",
    "\n",
    "***Cite***:  \n",
    "* Dan Wang, Hui Tang, Jian-Feng Liu, Shizhong Xu, Qin Zhang and Chao Ning. Rapid Epistatic Mixed Model Association Studies by Controlling Multiple Polygenic Effects. BioRxiv, 2020. doi: https://doi.org/10.1101/2020.03.05.976498  \n",
    "* Chao Ning, Dan Wang, Huimin Kang, Raphael Mrode, Lei Zhou, Shizhong Xu, Jian-Feng Liu. A rapid epistatic mixed-model association analysis by linear retransformations of genomic estimated values. Bioinformatics, 2018, 34(11): 1817-1825.  \n",
    "\n",
    "## Format of the input file.\n",
    "* Plink binary file including \\*.bed, \\*.bim and \\*.fam.  \n",
    "Missing genotypes are recommended to impute with Beagle or other softwares  \n",
    "\n",
    "* phenotypic file:  \n",
    "(1) Delimited by blanks or tabs;  \n",
    "(2) All individuals in the plink file must have phenotypic values. If no, please remove these individuals from the plink binary file;\n",
    "(3) The fisrt column is the family id and the second column is the individual id. The first two columns are the same to plink fam file, but order can be different;  \n",
    "(4) The last column is the phenotypic values. **Miss values are not allowed**;  \n",
    "(5) The covariates (including population means) are put before the phenotypic column. A column of 1’s must be contained.  \n",
    "An example phenotypic file with four covariates (population mean, sex, age, treatmeant or untreatmeant) is as follows:  \n",
    "12659\t14462\t1\t0\t126\t0\t0.58  \n",
    "12659\t14463\t1\t0\t91\t1\t0.39  \n",
    "12659\t14464\t1\t1\t126\t0\t0.37  \n",
    "12659\t14465\t1\t0\t91\t1\t0.9  \n",
    "12659\t14466\t1\t0\t91\t1\t0.84  \n",
    "12659\t14467\t1\t0\t91\t1\t0.61  \n",
    "12659\t14468\t1\t1\t91\t1\t0.84  \n",
    "  \n",
    "## Exhaustive additive by addtive epistatis  \n",
    "Data: Mouse data in directory of GMAT/examples/data/mouse    \n",
    "\n",
    "### Example 1: Include additive and additive by additive genomic relationship matrix\n",
    "#### 1. Exact test (for small data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_cpu.remma_epiAA_cpu import remma_epiAA_cpu\n",
    "from gmat.remma.annotation import annotation_snp_pos\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "pheno_file = 'pheno'\n",
    "bed_file = 'plink'\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "agmat_lst = agmat(bed_file, inv=False) # additive genomic relationship matrix\n",
    "dgmat_lst = dgmat_as(bed_file, inv=False) # dominace genomic relationship matrix\n",
    "\n",
    "# Step 2: Prepare the phenotypic vector (y), designed matrix for fixed effects (xmat) and designed matrix for random effects (zmat)\n",
    "y, xmat, zmat = design_matrix_wemai_multi_gmat(pheno_file, bed_file)\n",
    "\n",
    "# Step 3: Estimate the variances\n",
    "gmat_lst = [agmat_lst[0], agmat_lst[0]*agmat_lst[0]]  # agmat_lst[0]*agmat_lst[0] is the additive by additive genomic relationship matrix\n",
    "var_com_a_axa = wemai_multi_gmat(y, xmat, zmat, gmat_lst)\n",
    "print(var_com_a_ax)  # a list： [0] addtive variance; [1] additive by additive variance; [2] residual variance\n",
    "\n",
    "# Step 4: Test\n",
    "remma_epiAA_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_lst_0=None, p_cut=0.0001, out_file='remma_epiAA_cpu')\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "res_file = 'remma_epiAA_cpu'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### parallel \n",
    "The parameter snp_lst_0 means a list of snp index for the first SNP. Default value is None which means range(0, num_snp-1). For each SNP in the snp_lst_0, all SNPs after this SNP will pair with this SNP, i.e., the *i*th SNP will pari with range(i+1, num_snp).\n",
    "\n",
    "How to divide all SNPs equally? Taking three equal subset as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_df = pd.read_csv(bed_file + '.bim', header=None, sep='\\s+')\n",
    "num_snp = snp_df.shape[0]  # the number of snp\n",
    "num_snp_part = int(num_snp/6)  # Divide six parts\n",
    "part1 = list(range(0, num_snp_part)) + list(range(5*num_snp_part, num_snp-1))\n",
    "part2 = list(range(num_snp_part, 2*num_snp_part)) + list(range(4*num_snp_part, 5*num_snp_part))\n",
    "part3 = list(range(2*num_snp_part, 3*num_snp_part)) + list(range(3*num_snp_part, 4*num_snp_part))\n",
    "# submit following codes at the same time\n",
    "remma_epiAA_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_lst_0=part1, p_cut=0.0001, out_file='remma_epiAA_cpu.1')\n",
    "remma_epiAA_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_lst_0=part2, p_cut=0.0001, out_file='remma_epiAA_cpu.2')\n",
    "remma_epiAA_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_lst_0=part3, p_cut=0.0001, out_file='remma_epiAA_cpu.3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. approximate test (recommended for big data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import chi2\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.random_pair import random_pair\n",
    "from gmat.remma.remma_cpu.remma_epiAA_cpu import remma_epiAA_pair_cpu, remma_epiAA_eff_cpu_c\n",
    "from gmat.remma.annotation import annotation_snp_pos\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "pheno_file = 'pheno'\n",
    "bed_file = 'plink'\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "agmat_lst = agmat(bed_file, inv=False) # additive genomic relationship matrix\n",
    "dgmat_lst = dgmat_as(bed_file, inv=False) # dominace genomic relationship matrix\n",
    "\n",
    "# Step 2: Prepare the phenotypic vector (y), designed matrix for fixed effects (xmat) and designed matrix for random effects (zmat)\n",
    "y, xmat, zmat = design_matrix_wemai_multi_gmat(pheno_file, bed_file)\n",
    "\n",
    "# Step 3: Estimate the variances\n",
    "gmat_lst = [agmat_lst[0], agmat_lst[0]*agmat_lst[0]]  # agmat_lst[0]*agmat_lst[0] is the additive by additive genomic relationship matrix\n",
    "var_com_a_axa = wemai_multi_gmat(y, xmat, zmat, gmat_lst)\n",
    "print(var_com_a_axa)  # a list： [0] addtive variance; [1] additive by additive variance; [2] residual variance\n",
    "\n",
    "# Step 4: Randomly select 100,000 SNP pairs\n",
    "snp_df = pd.read_csv(bed_file + '.bim', header=None, sep='\\s+')\n",
    "num_snp = snp_df.shape[0]  # the number of snp\n",
    "random_pair(num_snp, out_file='random_pair', num_pair=100000, num_each_pair=5000)\n",
    "\n",
    "# step 5: Test these 100,000 SNP pairs\n",
    "# note: set p_cut=1 to save all the results\n",
    "remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_pair_file=\"random_pair\", max_test_pair=50000, p_cut=1, out_file='remma_epiAA_pair_cpu_random')\n",
    "\n",
    "# step 6: Calculate the median of variances for estimated epistatic SNP effects\n",
    "res_df = pd.read_csv('remma_epiAA_pair_cpu_random', header=0, sep='\\s+')\n",
    "print(np.median(res_df['p']))  # P value close to 0.5. It means type I error controlled well\n",
    "var_median = np.median(res_df['var'])  # median of variances for estimated epistatic SNP effects\n",
    "\n",
    "# step 7: Calculate the effect cut based on the p_cut\n",
    "p_cut = 1.0e-5\n",
    "chi_cut = chi2.isf(p_cut, 1)\n",
    "eff_cut = np.sqrt(chi_cut*var_median)\n",
    "\n",
    "# step 8: Screen the effects and select top SNP pairs based of effect cut\n",
    "remma_epiAA_eff_cpu_c(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_lst_0=None, eff_cut=eff_cut, out_file='remma_epiAA_eff_cpu_c')\n",
    "\n",
    "# Step 9: Calculate p values for top SNP pairs\n",
    "remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_pair_file=\"remma_epiAA_eff_cpu_c\", max_test_pair=50000, p_cut=1, out_file='remma_epiAA_pair_cpu_res')\n",
    "\n",
    "# Step 10: Select top SNPs and add the SNP position\n",
    "res_file = 'remma_epiAA_pair_cpu_res'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-7)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
