{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contents\n",
    "* [1 Contact](#1-Contact)\n",
    "* [2 Install](#2-Install)\n",
    "  * [2.1 Dependencies](#2.1-Dependencies)\n",
    "  * [2.2 Quick install](#2.2-Quick-install)\n",
    "  * [2.3 Detailed Package Install Instructions](#2.3-Detailed-Package-Install-Instructions)\n",
    "* [3 REMMAX function](#3-REMMAX-function)\n",
    "  * [3.1 Format of the input file](#3.1-Format-of-the-input-file)\n",
    "  * [3.2 Exhaustive additive by addtive epistatis](#3.2-Exhaustive-additive-by-addtive-epistatis)\n",
    "    * [3.2.1 Include additive and additive by additive genomic relationship matrix](#3.2.1-Include-additive-and-additive-by-additive-genomic-relationship-matrix)\n",
    "    * [3.2.2 Include additive, dominance and additive by additive genomic relationship matrix](#3.2.2-Include-additive,-dominance-and-additive-by-additive-genomic-relationship-matrix)\n",
    "    * [3.2.3 Include additive, dominance and three kinds of epistatic genomic relationship matrix](#3.2.3-Include-additive,-dominance-and-three-kinds-of-epistatic-genomic-relationship-matrix)\n",
    "  * [3.3 Exhaustive additive by dominance epistatis](#3.3-Exhaustive-additive-by-dominance-epistatis)\n",
    "  * [3.4 Exhaustive dominance by dominance epistatis](#3.4-Exhaustive-dominance-by-dominance-epistatis)\n",
    "  * [3.5 Exhaustive additive by additive epistatis with repeated measures](#3.5-Exhaustive-additive-by-additive-epistatis-with-repeated-measures)\n",
    "\n",
    "# 1 Contact\n",
    "Chao Ning  \n",
    "ningchao(at)sdau(dot)edu(dot)cn  \n",
    "ningchao91(at)gmail(dot)com  \n",
    "\n",
    "# 2 Install\n",
    "We will keep GMAT updated. Please uninstall older version to obtain the latest functions. The easiest uninstall way:  \n",
    "\\> pip uninstall gmat\n",
    "\n",
    "## 2.1 Dependencies\n",
    "* numpy>=1.16.0  \n",
    "* pandas>=0.19.0  \n",
    "* scipy>=1.1.1  \n",
    "* cffi>=1.12.0  \n",
    "* pandas_plink>=2.0.0  \n",
    "* tqdm>=4.43.0  \n",
    "\n",
    "We recommend using a Python distribution such as [Anaconda](https://www.anaconda.com/distribution/) (Python 3.7 version). This distribution can be used on Linux and Windows. It is the easiest way to get all the required package dependencies. \n",
    "\n",
    "## 2.2 Quick install\n",
    "\\> pip install gmat  \n",
    "\n",
    "## 2.3 Detailed Package Install Instructions\n",
    "(1) Install the dependent packages  \n",
    "(2) Go to the directory of GMAT and type  \n",
    "\\> python setup.py install  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3 REMMAX function  \n",
    "Rapid Epistatic Mixed Model Association Studies\n",
    "\n",
    "***Cite***:  \n",
    "* Dan Wang, Hui Tang, Jian-Feng Liu, Shizhong Xu, Qin Zhang and Chao Ning. Rapid Epistatic Mixed Model Association Studies by Controlling Multiple Polygenic Effects. *BioRxiv*, 2020. doi: https://doi.org/10.1101/2020.03.05.976498  \n",
    "* Chao Ning, Dan Wang, Huimin Kang, Raphael Mrode, Lei Zhou, Shizhong Xu, Jian-Feng Liu. A rapid epistatic mixed-model association analysis by linear retransformations of genomic estimated values. *Bioinformatics*, 2018, 34(11): 1817-1825.  \n",
    "\n",
    "## 3.1 Format of the input file\n",
    "* Plink binary file including \\*.bed, \\*.bim and \\*.fam.  \n",
    "Missing genotypes are recommended to impute with Beagle or other softwares, although they will be imputed according the frequency of occurrence.   \n",
    "\n",
    "* phenotypic file:  \n",
    "(1) Delimited by blanks or tabs;  \n",
    "(2) All individuals in the plink file must have phenotypic values. If no, please remove these individuals from the plink binary file;  \n",
    "(3) The fisrt column is the family id and the second column is the individual id. The first two columns are the same to plink fam file, but order can be different;  \n",
    "(4) The last column is the phenotypic values. **Miss values are not allowed**;  \n",
    "(5) The covariates (including population means) are put before the phenotypic column. A column of 1’s must be contained.  \n",
    "(6) Repeated mesures are allowed for individuals.  \n",
    "An example phenotypic file with four covariates (population mean, sex, age, treatmeant or untreatmeant) is as follows:  \n",
    "12659\t14462\t1\t0\t126\t0\t0.58  \n",
    "12659\t14463\t1\t0\t91\t1\t0.39  \n",
    "12659\t14464\t1\t1\t126\t0\t0.37  \n",
    "12659\t14465\t1\t0\t91\t1\t0.9  \n",
    "12659\t14466\t1\t0\t91\t1\t0.84  \n",
    "12659\t14467\t1\t0\t91\t1\t0.61  \n",
    "12659\t14468\t1\t1\t91\t1\t0.84  \n",
    "An example phenotypic file with repeated mesures is as follows:  \n",
    "0 01_01 1 -2.25383070574996  \n",
    "0 01_02 1 -1.88774565927855  \n",
    "0 01_03 1 2.4150679267528  \n",
    "0 01_03 1 -0.320697695608065  \n",
    "0 01_04 1 2.41743663901475  \n",
    "0 01_06 1 -0.634513668596019  \n",
    "0 01_06 1 -1.4489729404784  \n",
    "0 01_07 1 1.92328500921367  \n",
    "0 01_07 1 1.54547583777757  \n",
    "  \n",
    "## 3.2 Exhaustive additive by addtive epistatis  \n",
    "Data: Mouse data in directory of GMAT/examples/data/mouse    \n",
    "\n",
    "### 3.2.1 Include additive and additive by additive genomic relationship matrix\n",
    "#### (1)  Exact test (for small data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "gmat_lst = [ag, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_axa.txt')\n",
    "\n",
    "# Step 3: Test\n",
    "var_com = np.loadtxt('var_a_axa.txt')  # numpy array： [0] addtive variance; [1] additive by additive variance; [2] residual variance\n",
    "remma_epiAA(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, out_file='epiAA_a_axa')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA_a_axa'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Parallel exact test  (for small data)\n",
    "Analysis can be subdivided with remma_epiAA_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "gmat_lst = [ag, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_axa.txt')\n",
    "\n",
    "# Step 3: parallel test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "gmat_lst = [ag, ag*ag]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_parallel_a_axa')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "gmat_lst = [ag, ag*ag]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_parallel_a_axa')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "gmat_lst = [ag, ag*ag]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_parallel_a_axa')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_parallel_a_axa.*' with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_parallel_a_axa'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + \".merge\", 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos\n",
    "res_file = 'epiAA_parallel_a_axa.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (3) approximate test (recommended for big data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "gmat_lst = [ag, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_axa.txt')\n",
    "\n",
    "# Step 3: Approximate test\n",
    "var_com = np.loadtxt('var_a_axa.txt')  # numpy array： [0] addtive variance; [1] additive by additive variance; [2] residual variance\n",
    "remma_epiAA_approx(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, num_random_pair=100000, out_file='epiAA_approx_a_axa')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA_approx_a_axa'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (4) Parallel approximate test (recommended for big data)\n",
    "Analysis can be subdivided with remma_epiAA_approx_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "gmat_lst = [ag, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_axa.txt')\n",
    "\n",
    "# Step 3: parallel approximate test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "gmat_lst = [ag, ag*ag]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_approx_parallel_a_axa')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "gmat_lst = [ag, ag*ag]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_approx_parallel_a_axa')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "gmat_lst = [ag, ag*ag]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_approx_parallel_a_axa')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_approx_parallel_a_axa.*' \n",
    "# with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_approx_parallel_a_axa'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + '.merge', 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos                   \n",
    "res_file = 'epiAA_approx_parallel_a_axa.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2.2 Include additive, dominance and additive by additive genomic relationship matrix\n",
    "#### (1) Exact test (for small data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_d_axa.txt')\n",
    "\n",
    "# Step 3: Test\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')  # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; [3] residual variance\n",
    "remma_epiAA(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, out_file='epiAA_a_d_axa')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA_a_d_axa'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Parallel exact test (for small data)\n",
    "Analysis can be subdivided with remma_epiAA_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_d_axa.txt')\n",
    "\n",
    "# Step 3: parallel test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag] \n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_parallel_a_d_axa')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag] \n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_parallel_a_d_axa')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag] \n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_parallel_a_d_axa')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_parallel_a_d_axa.*' with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_parallel_a_d_axa'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + \".merge\", 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos\n",
    "res_file = 'epiAA_parallel_a_d_axa.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (3) approximate test (recommended for big data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_d_axa.txt')\n",
    "\n",
    "# Step 3: Approximate test\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')  # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; [3] residual variance\n",
    "remma_epiAA_approx(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, num_random_pair=100000, out_file='epiAA_approx_a_d_axa')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA_approx_a_d_axa'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (4) Parallel approximate test (recommended for big data)\n",
    "Analysis can be subdivided with remma_epiAA_approx_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var_a_d_axa.txt')\n",
    "\n",
    "# Step 3: parallel approximate test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_approx_parallel_a_d_axa')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_approx_parallel_a_d_axa')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var_a_d_axa.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_approx_parallel_a_d_axa')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_approx_parallel_a_d_axa.*' \n",
    "# with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_approx_parallel_a_d_axa'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + '.merge', 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos                   \n",
    "res_file = 'epiAA_approx_parallel_a_d_axa.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2.3 Include additive, dominance and three kinds of epistatic genomic relationship matrix\n",
    "additive, dominance, additive by additive, additive by dominance and dominance by dominance genomic relationship matrix   \n",
    "#### (1) Exact test (for small data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Test\n",
    "var_com = np.loadtxt('var.txt') # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; \n",
    "                                #               [3] additive by dominance variance; [4] dominance by dominance variance; [5] residual variance\n",
    "remma_epiAA(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, out_file='epiAA')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Parallel exact test (for small data)\n",
    "Analysis can be subdivided with remma_epiAA_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: parallel test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_parallel_a_d_axa.*' with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + \".merge\", 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos\n",
    "res_file = 'epiAA_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (3) approximate test (recommended for big data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Approximate test\n",
    "var_com = np.loadtxt('var.txt')  # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; [3] residual variance\n",
    "remma_epiAA_approx(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, num_random_pair=100000, out_file='epiAA_approx')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA_approx'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (4) Parallel approximate test (recommended for big data)\n",
    "Analysis can be subdivided with remma_epiAA_approx_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: parallel approximate test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_approx_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_approx_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_approx_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_approx_parallel_a_d_axa.*' \n",
    "# with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_approx_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + '.merge', 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos                   \n",
    "res_file = 'epiAA_approx_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.3 Exhaustive additive by dominance epistatis  \n",
    "Data: Mouse data in directory of GMAT/examples/data/mouse  \n",
    "Include additive, dominance, additive by additive, additive by dominance and dominance by dominance genomic relationship matrix  \n",
    "#### (1) Exact test (for small data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAD import remma_epiAD\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Test\n",
    "var_com = np.loadtxt('var.txt') # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; \n",
    "                                #               [3] additive by dominance variance; [4] dominance by dominance variance; [5] residual variance\n",
    "remma_epiAD(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, out_file='epiAD')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAD'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Parallel exact test (for small data)\n",
    "Analysis can be subdivided with remma_epiAD_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: parallel test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAD import remma_epiAD_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAD_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAD_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAD import remma_epiAD_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAD_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAD_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAD import remma_epiAD_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAD_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAD_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiAD_parallel.*' with the following codes.\n",
    "import os\n",
    "prefix = 'epiAD_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + \".merge\", 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos\n",
    "res_file = 'epiAD_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (3) approximate test (recommended for big data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAD import remma_epiAD_approx\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Approximate test\n",
    "var_com = np.loadtxt('var.txt')  # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; [3] residual variance\n",
    "remma_epiAD_approx(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, num_random_pair=100000, out_file='epiAD_approx')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAD_approx'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (4) Parallel approximate test (recommended for big data)\n",
    "Analysis can be subdivided with remma_epiAD_approx_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: parallel approximate test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAD import remma_epiAD_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAD_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAD_approx_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAD import remma_epiAD_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAD_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAD_approx_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAD import remma_epiAD_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAD_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAD_approx_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_approx_parallel.*' \n",
    "# with the following codes.\n",
    "import os\n",
    "prefix = 'epiAD_approx_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + '.merge', 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos                   \n",
    "res_file = 'epiAD_approx_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.4 Exhaustive dominance by dominance epistatis  \n",
    "Data: Mouse data in directory of GMAT/examples/data/mouse  \n",
    "Include additive, dominance, additive by additive, additive by dominance and dominance by dominance genomic relationship matrix   \n",
    "#### (1) Exact test (for small data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiDD import remma_epiDD\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Test\n",
    "var_com = np.loadtxt('var.txt') # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; \n",
    "                                #               [3] additive by dominance variance; [4] dominance by dominance variance; [5] residual variance\n",
    "remma_epiDD(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, out_file='epiDD')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiDD'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Parallel exact test (for small data)\n",
    "Analysis can be subdivided with remma_epiDD_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: parallel test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiDD import remma_epiDD_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiDD_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiDD_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiDD import remma_epiDD_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiDD_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiDD_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiDD import remma_epiDD_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg] \n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiDD_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiDD_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiDD_parallel.*' with the following codes.\n",
    "import os\n",
    "prefix = 'epiDD_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + \".merge\", 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos\n",
    "res_file = 'epiDD_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (3) approximate test (recommended for big data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiDD import remma_epiDD_approx\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file)  # additive genomic relationship matrix\n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Approximate test\n",
    "var_com = np.loadtxt('var.txt')  # numpy array： [0] addtive variance; [1] dominance variance; [2] additive by additive variance; [3] residual variance\n",
    "remma_epiDD_approx(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, num_random_pair=100000, out_file='epiDD_approx')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiDD_approx'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (4) Parallel approximate test (recommended for big data)\n",
    "Analysis can be subdivided with remma_epiDD_approx_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat, dgmat_as\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'plink'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "dgmat_as(bed_file)  # dominance genomic relationship matrix\n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'pheno'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')  # load the dominance genomic relationship matrix\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: parallel approximate test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiDD import remma_epiDD_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiDD_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiDD_approx_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiDD import remma_epiDD_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiDD_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiDD_approx_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiDD import remma_epiDD_approx_parallel\n",
    "bed_file = 'plink'\n",
    "pheno_file = 'pheno'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "dg = np.loadtxt(bed_file + '.dgrm_as0')\n",
    "gmat_lst = [ag, dg, ag*ag, ag*dg, dg*dg]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiDD_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiDD_approx_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiDD_approx_parallel.*' \n",
    "# with the following codes.\n",
    "import os\n",
    "prefix = 'epiDD_approx_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + '.merge', 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos                   \n",
    "res_file = 'epiDD_approx_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.5 Exhaustive additive by additive epistatis with repeated measures\n",
    "Data: Yeast data in directory of GMAT/examples/data/yeast   \n",
    "No heterozygous genotypes. No dominance effects.   \n",
    "  \n",
    "#### (1) Exact test (for small data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'CobaltChloride'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'CobaltChloride'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "pe = np.eye(ag.shape[0])  # identity matrix with dimension equal to the number of individuals. \n",
    "                          # to model the individual-specific error (permanent environmental effect)\n",
    "gmat_lst = [ag, ag*ag, pe]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Test\n",
    "var_com = np.loadtxt('var.txt')  # numpy array： [0] addtive variance; [1] additive by additive variance; \n",
    "                                             # [2] individual-specific error variance [3] residual variance\n",
    "remma_epiAA(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, out_file='epiAA')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Parallel exact test (for small data)\n",
    "Analysis can be subdivided with remma_epiAA_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.design_matrix import design_matrix_wemai_multi_gmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'CobaltChloride'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'CobaltChloride'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "pe = np.eye(ag.shape[0])  # identity matrix with dimension equal to the number of individuals. \n",
    "                          # to model the individual-specific error (permanent environmental effect)\n",
    "gmat_lst = [ag, ag*ag, pe]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "\n",
    "# Step 3: parallel test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'CobaltChloride'\n",
    "pheno_file = 'CobaltChloride'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "pe = np.eye(ag.shape[0]) \n",
    "gmat_lst = [ag, ag*ag, pe]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'CobaltChloride'\n",
    "pheno_file = 'CobaltChloride'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "pe = np.eye(ag.shape[0]) \n",
    "gmat_lst = [ag, ag*ag, pe]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_parallel\n",
    "bed_file = 'CobaltChloride'\n",
    "pheno_file = 'CobaltChloride'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "pe = np.eye(ag.shape[0]) \n",
    "gmat_lst = [ag, ag*ag, pe]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_parallel.*' with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + \".merge\", 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos\n",
    "res_file = 'epiAA_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=1.0e-5, dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (3) approximate test (recommended for big data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx\n",
    "from gmat.remma import annotation_snp_pos\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'CobaltChloride'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'CobaltChloride'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "pe = np.eye(ag.shape[0])  # identity matrix with dimension equal to the number of individuals. \n",
    "                          # to model the individual-specific error (permanent environmental effect)\n",
    "gmat_lst = [ag, ag*ag, pe]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: Approximate test\n",
    "var_com = np.loadtxt('var.txt')  # numpy array： [0] addtive variance; [1] additive by additive variance; \n",
    "                                             # [2] individual-specific error variance [3] residual variance\n",
    "remma_epiAA_approx(pheno_file, bed_file, gmat_lst, var_com, p_cut=1.0e-5, num_random_pair=100000, out_file='epiAA_approx')\n",
    "\n",
    "# Step 4: Select top SNPs and add the SNP position\n",
    "res_file = 'epiAA_approx'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=0.05*2/(28220*28219), dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (4) Parallel approximate test (recommended for big data)\n",
    "Analysis can be subdivided with remma_epiAA_approx_parallel and run parallelly on different machines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gmat.gmatrix import agmat\n",
    "from gmat.uvlmm.uvlmm_varcom import wemai_multi_gmat\n",
    "\n",
    "# Step 1: Calculate the genomic relationship matrix\n",
    "bed_file = 'CobaltChloride'  # the prefix for the plink binary file\n",
    "agmat(bed_file) \n",
    "\n",
    "# Step 2: Estimate the variances\n",
    "pheno_file = 'CobaltChloride'  # phenotypic file\n",
    "ag = np.loadtxt(bed_file + '.agrm0')  # load the additive genomic relationship matrix\n",
    "pe = np.eye(ag.shape[0])  # identity matrix with dimension equal to the number of individuals. \n",
    "                          # to model the individual-specific error (permanent environmental effect)\n",
    "gmat_lst = [ag, ag*ag, pe]  # ag*ag is the additive by additive genomic relationship matrix\n",
    "wemai_multi_gmat(pheno_file, bed_file, gmat_lst, out_file='var.txt')\n",
    "\n",
    "# Step 3: parallel approximate test. Write codes of thist step in separate scripts and run parallelly\n",
    "\n",
    "## parallel 1\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'CobaltChloride'\n",
    "pheno_file = 'CobaltChloride'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "pe = np.eye(ag.shape[0]) \n",
    "gmat_lst = [ag, ag*ag, pe]\n",
    "# parallel=[3, 1] means divide total tests into three parts and run part 1\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 1], p_cut=1.0e-5, out_file='epiAA_approx_parallel')\n",
    "\n",
    "## parallel 2\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'CobaltChloride'\n",
    "pheno_file = 'CobaltChloride'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "pe = np.eye(ag.shape[0]) \n",
    "gmat_lst = [ag, ag*ag, pe]\n",
    "# parallel=[3, 2] means divide total tests into three parts and run part 2\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 2], p_cut=1.0e-5, out_file='epiAA_approx_parallel')\n",
    "\n",
    "## parallel 3\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "import numpy as np\n",
    "from gmat.remma.remma_epiAA import remma_epiAA_approx_parallel\n",
    "bed_file = 'CobaltChloride'\n",
    "pheno_file = 'CobaltChloride'\n",
    "var_com = np.loadtxt('var.txt')\n",
    "ag = np.loadtxt(bed_file + '.agrm0')\n",
    "pe = np.eye(ag.shape[0]) \n",
    "gmat_lst = [ag, ag*ag, pe]\n",
    "# parallel=[3, 3] means divide total tests into three parts and run part 3\n",
    "remma_epiAA_approx_parallel(pheno_file, bed_file, gmat_lst, var_com, parallel=[3, 3], p_cut=1.0e-5, out_file='epiAA_approx_parallel')\n",
    "\n",
    "# Step 4: Merge files 'epiAA_approx_parallel.*' \n",
    "# with the following codes.\n",
    "import os\n",
    "prefix = 'epiAA_approx_parallel'\n",
    "parallel_num = 3  # the number of parallels\n",
    "with open(prefix + '.merge', 'w') as fout:\n",
    "    with open(prefix + '.1') as fin:\n",
    "        head_line = fin.readline()\n",
    "        fout.write(head_line)\n",
    "    for i in range(1, parallel_num+1):\n",
    "        with open(prefix + '.' + str(i)) as fin:\n",
    "            head_line = fin.readline()\n",
    "            for line in fin:\n",
    "                fout.write(line)\n",
    "        os.remove(prefix + '.' + str(i))\n",
    "\n",
    "# Step 5: Select top SNPs and add the SNP position\n",
    "from gmat.remma import annotation_snp_pos                   \n",
    "res_file = 'epiAA_approx_parallel.merge'  # result file\n",
    "annotation_snp_pos(res_file, bed_file, p_cut=0.05*2/(28220*28219), dis=0)  # p values < 1.0e-5 and the distance between SNP pairs > 0\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
